feat(azure-nvme-id): refactor and introduce unit tests

cjp256 opened the pull request:
Refactor main.c into several modules to support unit
testing and add unit test coverage for all methods.

No functional changes except a debug log or two.

CI changes:
- add Ubuntu 24.04 image to os matrix
- move dependency installation into build-deb.sh
- run unit tests in deb
- enforce clang-format
- enforce cppcheck
- enforce unit tests

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

Depends on #53

vmarcella commented on src/debug.h:
```diff
@@ -0,0 +1,26 @@
+extern bool debug;
+
+#define DEBUG_PRINTF(fmt, ...)                                                                                         \
+    do                                                                                                                 \
```
Why the do while loop?

cjp256 replied:
it gets compiled out, but it's a common pattern for macros like this to avoid problems during expansion.  https://hownot2code.wordpress.com/2016/12/05/do-while-0-in-macros/

vmarcella commented on src/debug.h:
```diff
@@ -0,0 +1,26 @@
+#include <stdbool.h>
+#include <stdio.h>
+
+extern bool debug;
```
NIT: Should this be named `DEBUG` since it's globally defined? Or does that only apply to global statics?

cjp256 replied:
ALLCAPS are typically macros in C.  Some code styles use g_ or global_ to help indicate as such.

vmarcella commented on cmake/tests.cmake:
```diff
@@ -0,0 +1,30 @@
+find_package(PkgConfig REQUIRED)
+pkg_check_modules(CMOCKA REQUIRED cmocka)
+
+function(add_test_executable test_name)
```
Nice, this a really cool solution for automating all the symbols that need to be wrapped for a test!

vmarcella reviewed (approved):
Don't have the best understanding of C or it's practices, but overall this looks good to me.

chewi reviewed (approved):
The changes look good at a glance, but I also tried running the tests as part of our Gentoo package, and that worked fine.

jepio commented on src/identify_disks.c:
```diff
@@ -0,0 +1,124 @@
+    {
+        struct nvme_controller ctrl;
+
+        strncpy(ctrl.name, namelist[i]->d_name, sizeof(ctrl.name));
```
would stick to snprintf here as well

cjp256 replied:
done!

jepio commented on src/identify_disks.c:
```diff
@@ -0,0 +1,124 @@
+    char vendor_id_path[MAX_PATH];
+    snprintf(vendor_id_path, sizeof(vendor_id_path), "%s/%s/device/vendor", SYS_CLASS_NVME_PATH, device_name);
+
+    char *vendor_id_string = read_file_as_string(vendor_id_path);
```
not that it matters much, but all the vendor udev files report 4096 as their length.

How about changing the utility to "read_hex_from_file" and doing:
```
FILE *f = fopen(argv[1], "r");
... // some error handling
long int x = 0;
int ret = fscanf(f, "%lx", &x);
if (ret != 1) {
   // return 0? return -1?
}
return x;
```

cjp256 replied:
I really like this idea but will do it in a follow up PR.

cjp256 replied:
actually, I removed util module and did it now! thanks!

cjp256 committed:
feat(azure-nvme-id): refactor and introduce unit tests

Refactor main.c into several modules to support unit
testing and add unit test coverage for all methods.

No functional changes except a debug log or two.

CI changes:
- add Ubuntu 24.04 image to os matrix
- move dependency installation into build-deb.sh
- run unit tests in deb
- enforce clang-format
- enforce cppcheck
- enforce unit tests

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

cjp256 committed:
use snprintf

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

cjp256 committed:
drop read_file_as_string() in favor of fscanf()

cjp256 committed:
build-deb: add pkg-config dep

Signed-off-by: Chris Patterson <cpatterson@microsoft.com>

cjp256 committed:
workflows/main:  drop duplicate install for clang-format

jeremycline commented on tests/identify_udev_tests.c:
```diff
@@ -0,0 +1,143 @@
+    assert_string_equal(capture_stderr(), "");
+}
+
+static void test_print_udev_key_values_for_vs_failure(void **state)
```
One other test case you might consider is weirder input like nothing but commas, multiple `=` characters, stuff like that.

jeremycline reviewed (approved):
Sorry for getting to this so late. I skimmed the refactor and didn't spot anything concerning (not that you can when skimming C). The test case coverage seems fairly comprehensive as well.
